### PR TITLE
Improve the domain check

### PR DESCRIPTION
**Pull request by mfairchild365**
URLs like `http://notallowed.com/?url=http://allowed.com/` were getting through.

### CODE REVIEW

File: src/UNL/VisitorChat/Util.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        }
+        
+        if (!isset($parts['host'])) {
+            //it isn't a valid host, don't allow
+            return false;
+        }
+
+        $regex = "";
+        foreach ($domains as $domain) {
+            $regex .= preg_quote($domain, ".-/") . "|";
```
Comment (kabel) [inline src/UNL/VisitorChat/Util.php]: I think your syntax is off for the `preq_quote` function. The second param should be the single character that is used as the pattern delimiter in `preg_match`. You might have better luck with:

```php
array_walk($domains, function(&$domain) {
    $domain = preg_quote($domain, "/");
});
$regex = implode($domains, "|");
```

File: src/UNL/VisitorChat/Util.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        }
+
+        $regex = trim($regex, "|");
+        if (preg_match("/" . $regex . "$/", $parts['host'])) {
```
Comment (kabel) [inline src/UNL/VisitorChat/Util.php]: The RegEx is close, but you'll want to wrap the `$regex` in parens `()` (group) to ensure the `$` (end constraint) is enforced for all of the domain combinations. Example of the problem: `/unl\.edu|nebraska\.edu$/` will match `unl.edu.phisherman.evil`. `/(unl\.edu|nebraska\.edu)$/`will not.

Comment (kabel) [inline src/UNL/VisitorChat/Util.php]: You might also want to add a constrained for better detecting sub-domains. What happens when someone uses the domain `iamevilprefixtest.com`

File: tests/Util/isAllowedDomain.phpt [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+$boolarray = Array(false => 'false', true => 'true');
+
+echo $boolarray[\UNL\VisitorChat\Util::isAllowedDomain('http://test.com', $allowed_domains)] . PHP_EOL;
+echo $boolarray[\UNL\VisitorChat\Util::isAllowedDomain('https://test.com', $allowed_domains)] . PHP_EOL;
+echo $boolarray[\UNL\VisitorChat\Util::isAllowedDomain('http://test.com/stuff.php?test', $allowed_domains)] . PHP_EOL;
+echo $boolarray[\UNL\VisitorChat\Util::isAllowedDomain('http://bad.com', $allowed_domains)] . PHP_EOL;
+echo $boolarray[\UNL\VisitorChat\Util::isAllowedDomain('http://bad.com/stuff.php?test', $allowed_domains)] . PHP_EOL;
+echo $boolarray[\UNL\VisitorChat\Util::isAllowedDomain('http://bad.com/stuff.php?test=http://test.com', $allowed_domains)] . PHP_EOL;
+echo $boolarray[\UNL\VisitorChat\Util::isAllowedDomain('http://subdomain.test.com', $allowed_domains)] . PHP_EOL;
```
Comment (kabel) [inline tests/Util/isAllowedDomain.phpt]: Beautiful tests. Add the following should be false:

* `test.com.bad.com`
* `iamnottest.edu`

**Comment by mfairchild365** [comment]
@kabel ready for another review

File: src/UNL/VisitorChat/Util.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        foreach ($domains as $domain) {
+            //Root domains are okay
+            $possibilities[] = '^' . preg_quote($domain, "/");
+            //Sub-domains are okay
+            $possibilities[] = '.*\.' . preg_quote($domain, "/");
+        }
+        
+        $regex = implode($possibilities, "|");
+        
+        $regex = "/(" . $regex . ")$/i";
```
Comment (kabel) [inline src/UNL/VisitorChat/Util.php]: I was just thinking that `$regex = "/^(.*\.)?(" . $regex . ")$/i"` might be easier than having to have to add a possibility for each domain.

**Pull request merged**